### PR TITLE
[IMP] website_forum: improve filters in forum

### PR DESCRIPTION
**Pull request by angv-odoo**
This commit improves filters in forum so that moderator can see all non-deleted posts on main page, use filters to move to flagged, closed or pending post lists, and reopen them through filters if need arises.

task-3349373

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by flch-odoo** [comment]
Replaces odoo-dev#2675.

File: addons/website_forum/controllers/website_forum.py [hunk 1]
```diff
@@ -334,7 +334,7 @@ def question_toggle_favorite(self, forum, question, **post):
 
     @http.route('/forum/<model("forum.forum"):forum>/question/<model("forum.post"):question>/ask_for_close', type='http', auth="user", methods=['POST'], website=True)
     def question_ask_for_close(self, forum, question, **post):
-        reasons = request.env['forum.post.reason'].search([('reason_type', '=', 'basic')])
+        reasons = request.env['forum.post.reason'].search(['|', ('reason_type', '=', 'basic'), ('reason_type', '=', 'offensive')])
```
Comment (std-odoo) [inline addons/website_forum/controllers/website_forum.py:337]: can you use expression.OR ?

File: addons/website_forum/controllers/website_forum.py [hunk 2]
```diff
@@ -366,12 +366,13 @@ def question_reopen(self, forum, question, **kwarg):
 
     @http.route('/forum/<model("forum.forum"):forum>/question/<model("forum.post"):question>/delete', type='http', auth="user", methods=['POST'], website=True)
     def question_delete(self, forum, question, **kwarg):
-        question.active = False
-        return request.redirect("/forum/%s" % slug(forum))
+        question.state = 'deleted'
+        return request.redirect("/forum/%s" % (slug(forum)))
```
Comment (std-odoo) [inline addons/website_forum/controllers/website_forum.py:370]: why this change ?

File: addons/website_forum/models/forum_post.py [hunk 3]
```diff
@@ -36,7 +36,7 @@ class Post(models.Model):
         [
             ('active', 'Active'), ('pending', 'Waiting Validation'),
             ('close', 'Closed'), ('offensive', 'Offensive'),
-            ('flagged', 'Flagged'),
+            ('flagged', 'Flagged'), ('deleted', 'Deleted')
```
Comment (std-odoo) [inline addons/website_forum/models/forum_post.py:39]: can you move it on a new line ?

File: addons/website_forum/models/forum_post.py [hunk 4]
```diff
@@ -476,10 +476,9 @@ def reopen(self):
         if any(post.parent_id or post.state != 'close' for post in self):
             return False
 
-        reason_offensive = self.env.ref('website_forum.reason_7')
-        reason_spam = self.env.ref('website_forum.reason_8')
+        reason_spam = self.env['forum.post.reason'].search([('name', '=', 'Spam or advertising')]).id
```
Comment (std-odoo) [inline addons/website_forum/models/forum_post.py:479]: I think you should use the XMLID instead (the name can be changed technically)

File: addons/website_forum/models/forum_post.py [hunk 5]
```diff
@@ -515,13 +514,20 @@ def close(self, reason_id):
...
-            'state': 'close',
-            'closed_uid': self._uid,
-            'closed_date': datetime.today().strftime(tools.DEFAULT_SERVER_DATETIME_FORMAT),
-            'closed_reason_id': reason_id,
-        })
+        if reason.reason_type == 'offensive':
+            self.write({
+                'state': 'offensive',
+                'closed_uid': self._uid,
+                'closed_date': datetime.today().strftime(tools.DEFAULT_SERVER_DATETIME_FORMAT),
```
Comment (std-odoo) [inline addons/website_forum/models/forum_post.py:521]: why do you convert it to string ?

File: addons/website_forum/models/forum_post.py [hunk 6]
```diff
@@ -515,13 +514,20 @@ def close(self, reason_id):
...
+                'closed_reason_id': reason_id,
+            })
+        else:
+            self.write({
+                'state': 'close',
+                'closed_uid': self._uid,
+                'closed_date': datetime.today().strftime(tools.DEFAULT_SERVER_DATETIME_FORMAT),
```
Comment (std-odoo) [inline addons/website_forum/models/forum_post.py:528]: same here

File: addons/website_forum/models/forum_post.py [hunk 7]
```diff
@@ -821,7 +826,15 @@ def _search_get_detail(self, website, order, options):
...
             domain = expression.AND([domain, [('has_validated_answer', '=', True)]])
         elif filters == 'unsolved':
-            domain = expression.AND([domain, [('has_validated_answer', '=', False)]])
+            domain += [('has_validated_answer', '=', False)]
+        elif filters == 'pending':
+            domain += [('state', '=', 'pending')]
+        elif filters == 'offensive':
+            domain += [('state', '=', 'offensive')]
+        elif filters == 'deleted':
+            domain += [('state', '=', 'deleted')]
```
Comment (std-odoo) [inline addons/website_forum/models/forum_post.py:835]: seems a bit redundant, can you do something like
```py
elif filters in ('pending', 'offensive', 'deleted'):
    domain += [('state', '=', filters)]

File: addons/website_forum/models/forum_post.py [hunk 8]
```diff
@@ -821,7 +826,15 @@ def _search_get_detail(self, website, order, options):
         elif filters == 'solved':
             domain = expression.AND([domain, [('has_validated_answer', '=', True)]])
         elif filters == 'unsolved':
-            domain = expression.AND([domain, [('has_validated_answer', '=', False)]])
+            domain += [('has_validated_answer', '=', False)]
```
Comment (std-odoo) [inline addons/website_forum/models/forum_post.py:829]: can you keep using expression.AND instead of `+=` ?

File: addons/website_forum/models/forum_post.py [hunk 9]
```diff
@@ -803,13 +806,15 @@ def _search_get_detail(self, website, order, options):
...
+        values = options.get('values') or False
         domain = website.website_domain()
-        domain = expression.AND([domain, [('state', '=', 'active'), ('can_view', '=', True)]])
+        domain = expression.AND([domain, [('can_view', '=', True)]])
         include_answers = options.get('include_answers', False)
         if not include_answers:
             domain = expression.AND([domain, [('parent_id', '=', False)]])
         forum = options.get('forum')
+        if values is False or values.get('user').karma < values.get('forum').karma_moderate:
+            domain += [('state', '!=', 'close'), ('state', '!=', 'pending')]
```
Comment (std-odoo) [inline addons/website_forum/models/forum_post.py:817]: same here, use expression.AND

File: addons/website_forum/models/forum_post.py [hunk 10]
```diff
@@ -803,13 +806,15 @@ def _search_get_detail(self, website, order, options):
...
-
+        values = options.get('values') or False
         domain = website.website_domain()
-        domain = expression.AND([domain, [('state', '=', 'active'), ('can_view', '=', True)]])
+        domain = expression.AND([domain, [('can_view', '=', True)]])
         include_answers = options.get('include_answers', False)
         if not include_answers:
             domain = expression.AND([domain, [('parent_id', '=', False)]])
         forum = options.get('forum')
+        if values is False or values.get('user').karma < values.get('forum').karma_moderate:
```
Comment (std-odoo) [inline addons/website_forum/models/forum_post.py:816]: why using `.get` and then `.karma` ?

if it's none it will crach anyway, so you can do
```suggestion
        if not values or values['user'].karma < values['forum'].karma_moderate:

**Review by std-odoo — COMMENTED** [review]
@angv-odoo Hi :) 

Some comments while passing, since it's a target and the freeze is near :) 

Cheers

File: addons/website_forum/models/forum_post.py (hunk 5 above)
Comment (angv-odoo) [reply addons/website_forum/models/forum_post.py:521]: I have not changed this part, 'closed_date' was already being converted to string before.

File: addons/website_forum/controllers/website_forum.py (hunk 2 above)
Comment (angv-odoo) [reply addons/website_forum/controllers/website_forum.py:370]: because we don't want the moderator to archive forums, instead put them in "deleted" state. we decided that archiving the forum would only be happening from back-end.

Comment (std-odoo) [reply addons/website_forum/controllers/website_forum.py:370]: I was asking for the `()` around slug :)

File: addons/website_forum/views/forum_forum_templates_post.xml [hunk 11]
```diff
@@ -240,6 +247,13 @@
                         <t t-set="karma" t-value="question.karma_close if not question.can_close else 0"/>
                         <t t-set="classes" t-value="'btn-info'"/>
                     </t>
+                    <t t-if="user.karma > question.karma_unlink" t-call="website_forum.link_button">
+                        <t t-set="url" t-value="'/forum/' + slug(forum) + '/question/' + slug(question) + '/delete'"/>
```
Comment (std-odoo) [inline addons/website_forum/views/forum_forum_templates_post.xml:251]: you can use t-valuef

File: addons/website_forum/views/forum_forum_templates_post.xml [hunk 12]
```diff
@@ -240,6 +247,13 @@
...
+                        <t t-set="label">Delete</t>
+                        <t t-set="icon" t-value="'fa-trash'"/>
```
Comment (std-odoo) [inline addons/website_forum/views/forum_forum_templates_post.xml:253]:
```suggestion
                        <t t-set="icon">fa-trash</t>

File: addons/website_forum/views/forum_forum_templates_post.xml [hunk 13]
```diff
@@ -240,6 +247,13 @@
...
+                        <t t-set="karma" t-value="not question.can_close and question.karma_close or 0"/>
+                        <t t-set="classes" t-value="'btn-danger'"/>
```
Comment (std-odoo) [inline addons/website_forum/views/forum_forum_templates_post.xml:255]: same here

File: addons/website_forum/views/forum_forum_templates_post.xml [hunk 14]
```diff
@@ -158,6 +164,7 @@
     </form>
 </template>
 
+
```
Comment (std-odoo) [inline addons/website_forum/views/forum_forum_templates_post.xml:167]: useless change

File: addons/website_forum/views/forum_forum_templates_layout.xml [hunk 15]
```diff
@@ -153,6 +154,32 @@
...
+                            <t t-if="user.karma>=forum.karma_moderate">
+                                <a t-attf-href="?#{ keep_query('search', 'sorting', 'my', filters='offensive') }"
+                                    class="dropdown-item">Offensive
+                                </a>
+                                <a t-attf-href="?#{ keep_query('search', 'sorting', 'my', filters='deleted') }"
+                                    class="dropdown-item">Deleted
+                                </a>
+                            </t>
+                            <t t-if="my=='mine' and forum.karma_moderate>=user.karma">
+                                <a t-attf-href="?#{ keep_query('search', 'sorting', 'my', filters='hidden') }"
```
Comment (std-odoo) [inline addons/website_forum/views/forum_forum_templates_layout.xml:179]: the t-if can be on the `<a/>` directly

**Review by std-odoo — COMMENTED** [review]
@angv-odoo Hi :) 

Next batch of comments :) Shouldn't you make a migration script about the "active field" change ?

Cheers

File: addons/website_forum/controllers/website_forum.py [hunk 16]
```diff
@@ -366,12 +366,13 @@ def question_reopen(self, forum, question, **kwarg):
 
     @http.route('/forum/<model("forum.forum"):forum>/question/<model("forum.post"):question>/delete', type='http', auth="user", methods=['POST'], website=True)
     def question_delete(self, forum, question, **kwarg):
-        question.active = False
+        question.state = 'deleted'
```
[the remark this answers is not in the document]
Comment (angv-odoo) [reply addons/website_forum/controllers/website_forum.py:369]: there can still be made archived records by internal users, we just don't want moderators to create those records from front-end. And I'm not sure how we would differentiate the two in older versions if we were to write a migration script...

[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/website_forum/controllers/website_forum.py:369]: IMO it something is "deleted" from a functionnal POV, it should be after your change as well

[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/website_forum/controllers/website_forum.py:369]: > there can still be made archived records by internal users

so it will be marked as deleted in the frontend right ? (so it's state should be deleted now)

[the remark this answers is not in the document]
Comment (angv-odoo) [reply addons/website_forum/controllers/website_forum.py:369]: So the idea is that: in the back-end user should be able to archive the question, in which case it will not change the state but have active field false. In front-end user will be able to change the state from active to offensive to deleted, but the active field will always stay True.

File: addons/website_forum/controllers/website_forum.py [hunk 17]
```diff
@@ -366,12 +366,13 @@ def question_reopen(self, forum, question, **kwarg):
...
         return request.redirect("/forum/%s" % slug(forum))
 
     @http.route('/forum/<model("forum.forum"):forum>/question/<model("forum.post"):question>/undelete', type='http', auth="user", methods=['POST'], website=True)
     def question_undelete(self, forum, question, **kwarg):
-        question.active = True
+        question.state = 'closed'
```
[the remark this answers is not in the document]
Comment (angv-odoo) [reply addons/website_forum/controllers/website_forum.py:374]: because when undeleted from front-end, we do not change the active field but the state

File: addons/website_forum/controllers/website_forum.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/website_forum/controllers/website_forum.py:369]: Yes ok, that'sq how it work now

But before your change, if a post is archived, it's marked as deleted

Now it won't be displayed as deleted on the frontend and it should

[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/website_forum/controllers/website_forum.py:369]: So you should do a migration script to keep the functionnal state of the post the same

File: addons/website_forum/models/forum_post.py [hunk 18]
```diff
@@ -491,15 +491,15 @@ def reopen(self):
                         karma *= 10
                 post.create_uid.sudo()._add_karma(karma * -1, post, _('Reopen a banned question'))
 
-        self.sudo().write({'state': 'active'})
+        self.sudo().write({'state': 'active', 'closed_reason_id': False})
```
[the remark this answers is not in the document]
Comment (angv-odoo) [reply addons/website_forum/models/forum_post.py:494]: when moving questions from closed to active, we should get rid of the closed_reason_id since it's not closed anymore

File: addons/website_forum/controllers/website_forum.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/website_forum/controllers/website_forum.py:369]: Just do the test, make a db in master, make a screenshot of the frontend (of a deleted post), then upgrade

It should be the exact same after the upgrade

File: addons/website_forum/views/forum_forum_templates_post.xml (hunk 12 above)
Comment (flch-odoo) [reply addons/website_forum/views/forum_forum_templates_post.xml]: Won't this make it translatable?

File: addons/website_forum/views/forum_forum_templates_post.xml (hunk 13 above)
Comment (flch-odoo) [reply addons/website_forum/views/forum_forum_templates_post.xml:254]: same too

File: addons/website_forum/controllers/website_forum.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (flch-odoo) [reply addons/website_forum/controllers/website_forum.py:369]: I think it's part of the changes we want here.
> In front-end user will be able to change the state from active to offensive to deleted, but the active field will always stay True.

this is not how it works now, but how it should.
An archived post shouldn't show at all on the front-end.

**Review by flch-odoo — COMMENTED** [review]
just passing through ;)

File: addons/website_forum/controllers/website_forum.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/website_forum/controllers/website_forum.py:369]: @flch-odoo So it's ok if someone delete a post in 16.0 (so it's archived), it's displayed as deleted on the frontend, and after the 17.0 upgrade is not displayed at all and the state is still the old one ?

> I think it's part of the changes we want here.

Yes I know, the point is to keep the functionnal state of the posts after the upgrade

[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/website_forum/controllers/website_forum.py:369]: to discuss tomorrow at GR :)

File: addons/website_forum/views/forum_forum_templates_post.xml (hunk 12 above)
Comment (std-odoo) [reply addons/website_forum/views/forum_forum_templates_post.xml:252]: ahyes you are right, we can use t-valuef instead of t-value then :)

**Comment by flch-odoo** [comment]
Superseeded by #138181.
